packages/contracts-bedrock/scripts/invariant-doc-gen.ts

```diff
@@ -0,0 +1,168 @@
+}
+
+// Represents the documentation of an invariant
+type InvariantDoc = {
+  header?: string
+  desc?: string
+  lineNo?: number
+}
+
+const writtenFiles = []
+
+/**
+ * Lazy-parses all test files in the `contracts/test/invariants` directory to generate documentation
+ * on all invariant tests.
+ */
+const docGen = (dir: string): void => {
+  // Grab all files within the invariants test dir
+  const files = fs.readdirSync(dir)
+
+  // Array to store all found invariant documentation comments.
+  const docs: Contract[] = []
+
+  for (const fileName of files) {
+    // Read the contents of the invariant test file.
+    const fileContents = fs.readFileSync(`${dir}/${fileName}`).toString()
+
+    // Split the file into individual lines and trim whitespace.
+    const lines = fileContents.split('\n').map((line: string) => line.trim())
+
+    // Create an object to store all invariant test docs for the current contract
+    const isEchidna = fileName.startsWith('Fuzz')
+    const name = isEchidna
+      ? fileName.replace('Fuzz', '').replace('.sol', '')
+      : fileName.replace('.t.sol', '')
+    const contract: Contract = { name, fileName, isEchidna, docs: [] }
+
+    let currentDoc: InvariantDoc
+
+    // Loop through all lines to find comments.
+    for (let i = 0; i < lines.length; i++) {
+      let line = lines[i]
+
+      if (line.startsWith('/**')) {
+        // We are at the beginning of a new doc comment. Reset the `currentDoc`.
+        currentDoc = {}
+
+        // Move on to the next line
+        line = lines[++i]
+
+        // We have an invariant doc
+        if (line.startsWith(`* ${NATSPEC_INV}`)) {
+          // Assign the header of the invariant doc.
+          // TODO: Handle ambiguous case for `INVARIANT: ` prefix.
+          // TODO: Handle multi-line headers.
+          currentDoc = {
+            header: line.replace(`* ${NATSPEC_INV}`, '').trim(),
+            desc: '',
+          }
+
+          // If the header is multi-line, continue appending to the `currentDoc`'s header.
+          while (BLOCK_COMMENT_HEADER_REGEX.test((line = lines[++i]))) {
+            currentDoc.header += ` ${line
+              .replace(BLOCK_COMMENT_PREFIX_REGEX, '')
+              .trim()}`
+          }
+
+          // Process the description
+          while ((line = lines[++i]).startsWith('*')) {
+            line = line.replace(BLOCK_COMMENT_PREFIX_REGEX, '').trim()
+
+            // If the line has any contents, insert it into the desc.
+            // Otherwise, consider it a linebreak.
+            currentDoc.desc += line.length > 0 ? `${line} ` : '\n'
+          }
+
+          // Set the line number of the test
+          currentDoc.lineNo = i + 1
+
+          // Add the doc to the contract
+          contract.docs.push(currentDoc)
+        }
+      }
+    }
+
+    // Add the contract to the array of docs
+    docs.push(contract)
+  }
+
+  for (const contract of docs) {
+    const fileName = `${BASE_DOCS_DIR}/${contract.name}.md`
```

Comment:
Would it be annoying to use fully qualified names here? Ie `<filename>:<contractname>` or i'm fine with just enforcing that contract names must be unique

Reply:
i think colons in filenames will break windows

Reply:
Ah yes- so the `contract`'s `name` is found through the file name containing the tests, which assumes that for `forge` invariants, the file is named `ContractName.t.sol` in the `contracts/test/invariants` dir, and for `echidna`, named `FuzzContractName.sol` in the `contracts/echidna` dir.

For example, w/ `forge` invariants, each invariant test file will likely have many contracts within it due to differing `setUp`s (see: [Portal Invariants](https://github.com/ethereum-optimism/optimism/blob/97b1f605d3770ab603fd27e7f08c8d313dff95fc/packages/contracts-bedrock/contracts/test/invariants/OptimismPortal.t.sol)), so separating these into different files would probably add a good deal of clutter. LMK if I'm misunderstanding 😄